sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -106,7 +106,10 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
   }
 
   protected lazy val (buildKeys, streamedKeys) = {
-    require(leftKeys.map(_.dataType) == rightKeys.map(_.dataType),
+    require(
+      leftKeys.map(_.dataType)
+        .zip(rightKeys.map(_.dataType))
```

Comment:
`zip` allows different length, we should check it explicitly.

Reply:
ok